backend/gsr_booking/api_wrapper.py

```diff
@@ -532,38 +309,239 @@ def get_affiliation(self, email):
+        gsr = get_object_or_404(GSR, gid=gid)
+
+        # select a random user from the group if booking wharton gsr
+        if gsr.kind == GSR.KIND_WHARTON and group is not None:
+            wharton_members = group.memberships.filter(is_wharton=True)
+            if (n := wharton_members.count()) == 0:
+                raise APIError("Error: Non Wharton cannot book Wharton GSR")
+            user = wharton_members[randint(0, n - 1)].user
+
+        rooms = (
+            self.WBW.get_availability(lid, start, end, user)
+            if gsr.kind == GSR.KIND_WHARTON
+            else self.LBW.get_availability(gid, start, end, user)
         )
-        total_minutes = 0
-        for reservation in reservations:
-            # accumulates total minutes over all reservations
-            total_minutes += int((reservation.end.timestamp() - reservation.start.timestamp()) / 60)
-        # 120 minutes per day
-        return {CreditType.LIBCAL.value: 120 - total_minutes}
+        return {"name": gsr.name, "gid": gsr.gid, "rooms": rooms}
+
+    def get_reservations(self, user, group=None):
+        q = Q(user=user) | Q(reservation__creator=user) if group else Q(user=user)
+        bookings = GSRBooking.objects.filter(
+            q, is_cancelled=False, end__gte=timezone.localtime()
+        ).prefetch_related(Prefetch("reservation"))
+
+        if group:
+            ret = []
+            for booking in bookings:
+                data = GSRBookingSerializer(booking).data
+                if booking.reservation.creator == user:
+                    data["room_name"] = f"[Me] {data['room_name']}"
+                else:
+                    data["room_name"] = f"[{group.name}] {data['room_name']}"
+                ret.append(data)
+        else:
+            ret = GSRBookingSerializer(bookings, many=True).data
+
+        # deal with bookings made directly through wharton (not us)
+        try:
+            wharton_bookings = self.WBW.get_reservations(user)
+        except APIError:
+            return ret
+
+        if len(wharton_bookings) == 0:
+            return ret
+
+        booking_ids = set([booking["booking_id"] for booking in ret])
+        wharton_bookings = [
+            booking for booking in wharton_bookings if booking["booking_id"] not in booking_ids
+        ]
+        if len(wharton_bookings) == 0:
+            return ret
+
+        wharton_gsr_datas = {
+            gsr.gid: GSRSerializer(gsr).data for gsr in GSR.objects.filter(kind=GSR.KIND_WHARTON)
+        }
+        for booking in wharton_bookings:
+            booking["gsr"] = wharton_gsr_datas[booking["gid"]]
+            del booking["gid"]
+            ret.append(booking)
+        return ret
+
+    def check_credits(self, user):
+        pass  # seems like its unused on the frontend
+
+
+# initialize singletons
+WhartonGSRBooker = WhartonBookingWrapper()
+LibCalGSRBooker = LibCalBookingWrapper()
+GSRBooker = BookingHandler(WhartonGSRBooker, LibCalGSRBooker)
```

Comment:
can we do something like this https://www.geeksforgeeks.org/singleton-pattern-in-python-a-complete-guide/#

Reply:
isn't this the Module-level Singleton described there? Are you saying a different type of singleton? I think the module level is the easiest/cleanest for our purposes